### PR TITLE
Support for video capture via the MediaRecorder API

### PR DESCRIPTION
**Pull request by jeffposnick**
R: @ebidel @addyosmani 

This is code for a 2.0.0 release of `<google-youtube-upload>`. Like the previous version, you can upload a local video file, but this version will use the new [MediaRecorder API](https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder) in supported browsers (Chrome and Firefox at the moment) to allow for uploads captured from your local camera stream.

There's a live version of the demo at https://jeffy.info/google-youtube-upload/components/google-youtube-upload/demo/

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -17,16 +16,24 @@
...
-    "google-signin": "GoogleWebComponents/google-signin#^1.0.0",
-    "cors-upload-sample": "googledrive/cors-upload-sample#master"
+    "cors-upload-sample": "googledrive/cors-upload-sample#master",
+    "google-apis": "GoogleWebComponents/google-apis#^1.1.5",
+    "google-signin": "GoogleWebComponents/google-signin#^1.3.2",
+    "iron-flex-layout": "PolymerElements/iron-flex-layout#~1.3.1",
+    "iron-iconset-svg": "PolymerElements/iron-iconset-svg#^1.0.9",
+    "iron-pages": "PolymerElements/iron-pages#^1.0.7",
+    "md-gum-polyfill": "mozdevs/mediaDevices-getUserMedia-polyfill#master",
+    "paper-fab": "PolymerElements/paper-fab#~1.2.0",
```
Comment (ebidel) [inline bower.json]: `~` -> `^` for all of these.

File: demo/index.html [hunk 2]
```diff
@@ -5,14 +5,145 @@
...
+        width: 400px;
+        height: 300px;
       }
     </style>
   </head>
+
   <body>
-    <youtube-upload-demo-element></youtube-upload-demo-element>
+    <template is="dom-bind" id="app">
+      <p>
```
Comment (ebidel) [inline demo/index.html]: These `<p>` can be outside the dom-bind

File: demo/index.html [hunk 3]
```diff
@@ -5,14 +5,145 @@
...
+        var error;
+
+        switch(e.detail.uploadStatus) {
+          case 'failed':
+            error = e.detail.failureReason || 'unknown error';
+          break;
+
+          case 'rejected':
+            error = e.detail.rejectionReason || 'unknown error';
+          break;
```
Comment (ebidel) [inline demo/index.html]: I usually line `break` up with the block's code

File: google-youtube-upload.html [hunk 4]
```diff
@@ -21,447 +30,687 @@
...
+       * [the steps](https://developers.google.com/console/help/new/#generatingoauth2)
+       * for registering for OAuth 2, ensure that the
+       * [YouTube Data API v3](https://developers.google.com/youtube/registering_an_application)
+       * is enabled for your API project, and ensure that the JavaScript Origin
+       * is set to the domain hosting the page on which
+       * you'll be using this element.
        *
-       * @event youtube-upload-start
-       * @param {Object} e Event parameters.
+       * @attribute clientId
```
Comment (ebidel) [inline google-youtube-upload.html]: you no longer need these doc strings. i-c-p will pick up the type/value/name

File: google-youtube-upload.html [hunk 5]
```diff
@@ -21,447 +30,687 @@
...
-       * One or more `youtube-processing-poll` events will then be fired after this event,
-       * followed by either a `youtube-processing-complete` or `youtube-processing-fail`.
+       * @attribute tags
+       * @type Array.<string>
+       * @default ['google-youtube-upload']
+       */
+      tags: {
+        type: Array,
+        value: function() {
+          return ['google-youtube-upload']
```
Comment (ebidel) [inline google-youtube-upload.html]: `return ['google-youtube-upload'];`

File: google-youtube-upload.html [hunk 6]
```diff
@@ -21,447 +30,687 @@
...
+      this._uploadStartTime = Date.now();
+      uploader.upload();
+
+      this._transitionToThanks();
+    },
+
+    // Private helper methods.
+    _initializeMediaRecorder: function() {
+      if (this._mediaRecorder) {
+        return Promise.resolve();
```
Comment (ebidel) [inline google-youtube-upload.html]: should we load the promise-poyfill?

File: google-youtube-upload.html [hunk 7]
```diff
@@ -21,447 +30,687 @@
...
+              };
             }
           }.bind(this)
         });
       }
-    });
-  })();
+    },
+
+    _pollForVideoStatus: function() {
```
Comment (ebidel) [inline google-youtube-upload.html]: too bad the API doesn't have a better approach than polling. I remember the same type of thing with the drive zip download endpoint :\

**Comment by ebidel** [comment]
Sweet demo! Reviewed.

Trying to record a video slows down the FPS to a crawl in Chrome stable. Have you see that? The captured video is janky.

Another UX thing is when I click the upload video icon, there's no way to cancel out of it (e.g. switch back to picking the record video).

File: google-youtube-upload.html (hunk 6 above)
Comment (jeffposnick) [inline google-youtube-upload.html]: Given that the MediaRecorder API needs to be implemented in order for this line to execute, I don't think there's a real risk of it running on a browser that doesn't support Promises.

File: demo/index.html (hunk 2 above)
Comment (jeffposnick) [inline demo/index.html]: Moved.

File: bower.json (hunk 1 above)
Comment (jeffposnick) [inline bower.json]: Changed.

File: demo/index.html (hunk 3 above)
Comment (jeffposnick) [inline demo/index.html]: Switch()ed. (Hah.)

File: google-youtube-upload.html (hunk 5 above)
Comment (jeffposnick) [inline google-youtube-upload.html]: Done.

File: google-youtube-upload.html (hunk 4 above)
Comment (jeffposnick) [inline google-youtube-upload.html]: Oh, that's a nice change. I'll remove them.

File: google-youtube-upload.html (hunk 7 above)
Comment (jeffposnick) [inline google-youtube-upload.html]: We had a provisional way pushing API responses to a client, but it never actually shipped: https://www.youtube.com/watch?v=NlZZghBnfdM

**Comment by jeffposnick** [comment]
I have seen the Chrome Stable slowdown on OS X, yup. It's there is no slowdown in Canary M51, so I assumed it was a case where there was a less-than-optimal MediaRecorder API implemented in M49, which subsequently got more efficient. CC: @samdutton in case he knows.

The UX is what happens when I end up designing interfaces :-/ You can get back to the login/logout screen by tapping on your profile icon. Happy to implement via some other means if you have a suggestion.

**Comment by ebidel** [comment]
Chrome 51 is much better!

Didn't even try the profile icon. A material design back arrow might be nice, but no big deal if you don't want to add it.

LGTM

**Comment by jeffposnick** [comment]
When YouTube rolled out multi-account support and G+ Pages tied to channels, we encouraged developers to be super-explicit about which YouTube channel you were uploading into. (Kind of like making sure which Tweetdeck account you're acting as when you're composing a tweet.) So I wanted to keep the profile picture visible, and that plus the back arrow seems like a lot.

But at the same time, I'm not one for UI design, so if there's a way to fit them all in without overwhelming things, I can switch it up. Maybe in a future revision.

**Comment by samdutton** [comment]
+ Miguel

**Comment by miguelao** [comment]
@jeffposnick smartly composing a MediaStream can be done by drawing the necessary bits and pieces on a `<canvas>`, then extracting a `MediaStream` out of it with `<canvas>.captureStream()`; this `MediaStream` can be recorded (like in [this LayoutTest](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/LayoutTests/fast/mediacapturefromelement/CanvasCaptureMediaStream-framerate-0.html&sq=package:chromium&type=cs&l=53)). 

Canvas captureStream is working in Firefox and behind a flag in Chrome (ITS behind finalized). @uysalere knows more about the particularities that might arise, and, just to be complete, here's [the link](https://cdn.rawgit.com/uysalere/js-demos/master/intro.html
) (again) with the canvas capture demos.

**Comment by jeffposnick** [comment]
@miguelao, could you explain what the benefit would be of using the `<canvas>` approach vs. what the code is doing now? Is it a question of memory efficiency? Something else?

(Sorry, I'm new to most of the Media APIs!)

**Comment by jeffposnick** [comment]
I followed up with @miguelao off-thread, and while some of the composition options (like WebGL filters!) might be cool for a future release, the current basic implementation should work. I'm going to merge and tag a new release.

**Pull request merged**